src/learn/pathogens/index.rst

```diff
@@ -1,13 +1,14 @@
-
 ======================================
 Pathogens
 ======================================
 
-Learn about some of the pathogens analyzed by the Nextstrain team.
+The Nextstrain team maintains `analyses of several selected pathogens <pathogens_>`_, such as SARS-CoV-2_ and influenza_, to show pathogen evolution and epidemic spread.
+We continually update our results as new data is made publicly available by re-running our :term:`core builds <core build>`.
```

Comment:
Oh, hmm! When writing this text, I tried to get it right by closely reading the existing glossary, so I guess I'm still a bit fuzzy on _workflow_ vs. _build_ (the noun). :joy: My reading was:

- _workflow_ is the generic, parameterized thing
- _build_ is a specific combination of workflow + parameter values
- _dataset_ is (one of) the result/output of a _build_

So my thinking when writing the above was that we weren't re-running just our core workflows, but our core builds (workflows + specific params).

But it sounds like what you're saying is you envisioned the _workflow_ as **containing** the _build_? Almost like a base/sub-class relationship. This is a bit different from my thinking, which was the _workflow_ was like a class which you then instantiate with parameters to get a _build_. (Or alternatively, a _workflow_ is the mathematical model and a _build_ is the model + values for its free parameters).

(This all starts to feel like very finely split hairs, esp. when we ourselves are confused. I hope we can further clarify terms here.)

Reply:
> But it sounds like what you're saying is you envisioned the _workflow_ as **containing** the _build_?

Yeah, that's what's made sense in my head when I added [glossary](https://docs.nextstrain.org/en/latest/reference/glossary.html) and updated [Parts of a whole](https://docs.nextstrain.org/en/latest/learn/parts.html) with this figure:

<img width="500" alt="workflow-vs-build" src="https://user-images.githubusercontent.com/13424970/158482073-80261520-42d8-4a63-ba7e-3aa46ed101b4.png">

More importantly, a workflow can contain more than 1 build - like how we often name the config file "builds.yaml" and there is a `builds:` section within that.

This new terminology defers from historical usage on the team since the "build" term has been overloaded ([Slack message](https://bedfordlab.slack.com/archives/C02Q4Q5S85A/p1640074769005000?thread_ts=1639761464.002400&cid=C02Q4Q5S85A)). It still makes sense when we say "run the Zika build" since there is actually just one build that gets run in the Zika workflow. But when we say "run the ncov build" with the new terminology, that doesn't make sense since ncov is a _workflow_ which runs the ncov/global build and regional builds altogether.

Open to any suggestions to improve definitions, especially from more experienced folks.

Reply:
Your idea of _workflows + specific params_ is new though, I haven't considered that with existing terminology. If thinking of `ncov` as a _workflow_ implemented with Snakemake (conveniently a _workflow manager_), then _workflows + specific params_ would be what's described in [Configuration | Snakemake docs](https://snakemake.readthedocs.io/en/v5.1.4/snakefiles/configuration.html). So just a workflow with a set of configuration files?

Reply:
Right, config is "just another" workflow parameter (albeit a major, complex one).

Workflows containing builds feels like the wrong relationship to me. To give another example of something that feels more right to me: builds are a higher-level abstraction over a workflow, they're a tuple (workflow, parameters) which given inputs produce output datasets. (There's some wiggle room there, e.g. inputs are just another parameter or vice versa, but I think it doesn't matter much.)

As you say, we've been loose with _build_, _workflow_, etc. in the past. It's definitely good to keep trying to nail these down and be more specific and consistent in our writing.

Separately, but related, I don't think we want to use/define _workflow_ differently than the common usage (e.g. a Snakemake workflow, a Nextflow workflow, a WDL workflow, etc). Doing so would be making a choice to overload it, which seems counterproductive to clarity.

Reply:
> But when we say "run the ncov build" with the new terminology, that doesn't make sense since ncov is a _workflow_ which runs the ncov/global build and regional builds altogether.

Nod, though it makes sense again with the current glossary terms if you say "run the ncov build**s**"?

Reply:
Oh, wow. I haven't thought very deeply about these particulars. Without thinking it through, I've been largely using "workflow" and "build" interchangeably. I had (until just now) considered it just fine to say:

- "I just ran the ncov build to update regional datasets"
- "I ran the seasonal flu build to update datasets for H3N2, H1N1pdm, etc..."

I was just fine with pathogen repo =  "build" which when run may produce 1 or more datasets, ie Auspice JSONs.

I see however that this doesn't match our usage of "builds" here: https://github.com/nextstrain/ncov/blob/master/nextstrain_profiles/nextstrain-gisaid/builds.yaml#L32, which clearly delineates each workflow target as a separate "build".

So, I don't know what's best. I do think it would be nice if we had:
- fewer concepts where possible
- a simple word for the actual thing that is the pathogen repo

Reply:
> I see however that this doesn't match our usage of "builds" here: https://github.com/nextstrain/ncov/blob/master/nextstrain_profiles/nextstrain-gisaid/builds.yaml#L32, which clearly delineates each workflow target as a separate "build".

This (and the `builds.yaml` filename itself) is the major thing that prompted me to define workflow vs. build like I've done.

If it were instead `datasets.yaml` with a `datasets:` section, that would eliminate the need for the "build" term I've defined. This also makes sense in the UI where we have a "Datasets" selector for these entries.

Reply:
This makes sense to me. I think it would be my preference to drop from "builds", "workflows" and "datasets" to just "builds" and "datasets" (with "workflow" being synonymous with "build").

But would want to make sure this made sense to the larger group as well.

Reply:
> don’t think we want to use/define *workflow* differently than the common usage

> fewer concepts where possible

These are two guiding lights we should follow I think 💯 

---

To me, {workflow} vs {workflow + parameters} is splitting hairs a bit, and I think it would be too confusing to expect people to understand that’s what we mean by workflow vs build (no matter how clearly we define it in the glossary!). 

> If it were instead `datasets.yaml` with a `datasets:` section

This is appealing to me as it essentially frees up the word “build”. I think we are so accustomed to using this word in a context-dependent manner that defining it in a narrow context may not be possible. For our documentation we should use the most precise language we can, but for day-to-day speech I think saying “re-run the ncov builds” or “here’s the new north-america build”  is just fine.

Is there any common concept that can’t be communicated without workflow + datasets?

Reply:
My thoughts:

I don't necessarily want to make workflow/build formally distinct things in our terminology. My description of build as workflow+parameters was based on my reading/understanding of the current glossary (a reading which was different from Victor's intent! :upside_down_face:). It wasn't an alternative I was proposing.

That said, I do think workflow/build will keep some distinction informally because our use of build is so fluid/context-dependent, as @jameshadfield notes. So I don't think they're exactly synonymous (as in @trvrb's preference). For example, I think it's reasonable to say something like, "The ncov build has a Snakemake workflow and WDL workflow."  Or (hypothetically), "The ncov build contains two workflows: one for ingest and data provisioning and another for producing Nextstrain datasets."  We also, again as James notes, use build informally to mean dataset at times.

If we have to pin down build, what @trvrb wrote feels closest to the right definition to me:

> I was just fine with pathogen repo = "build" which when run may produce 1 or more datasets, ie Auspice JSONs.

as this encompasses both the workflow(s) and produced datasets.

I don't think ncov's builds.yaml should be called builds.yaml or datasets.yaml.  It's workflow config/parameters that contains a lot more than just the `builds:` key.  I think a name like config.yaml (but this is used by Snakemake's profile support) or parameters.yaml (what we use for defaults) would be more accurate.  My preference would be the latter given the conflict with Snakemake profile config vs. Snakemake workflow config.

Renaming builds.yaml to something else seems like it would be quite disruptive at this point.  Even if we make the change backwards compat, people will be confused when the terminology changes.  Build is used pretty pervasively in the ncov docs now.

We can use our current "best" names for things in other pathogen repos, but maybe should leave ncov a bit more stable.

I think the ncov `builds:` key would be better off being named `outputs:`, mirroring `inputs:`.  `datasets:` is ok too, but feels too specific/restrictive in this context.

Renaming the `builds:` key is maybe less disruptive than the file, but really not sure about that.

Reply:
Thanks so much for the thoughts here Tom. I agree with basically everything here. Though I'm okay not worrying too much about `builds.yaml` in `ncov`. I just didn't want to feel like we're running in circles to accommodate in our terminology a file name that was chosen in the thick of things in early 2020.

I found this particularly clarifying with regards to (necessary) distinctions between "build" and "workflow":
> "The ncov build has a Snakemake workflow and WDL workflow."
Many pathogen builds (ie repos) will have a single canonical workflow that happens to be in Snakemake. But the seasonal flu build (ie repo) has a workflow that's run for the live nextstrain.org site and a separate workflow that's run for private.nextflu.org.

Reply:
> ncov `builds:` key would be better off being named `outputs:`

Let's do this, at least for future workflows we develop!

Reply:
Thanks for the discussion here! It seems like `builds.yaml` and `builds:` is a one-off that's not straightforward to change. Maybe we can ignore it in defining terminology, and note in the `ncov` docs that the usage of "builds" in this file isn't reflective of standard terminology.

For future workflows, I like the idea of `outputs:` and a filename of something like `config.yml` (since it is used for `--configfile`). Conflict with Snakemake profile support should be less of an issue once we drop usage of profiles.

Reply:
Created a new issue #100 so we can continue discussion outside of this PR.

Reply:
A couple of belated thoughts:

Renaming `builds.yaml` to anything else is much less disruptive than renaming the `builds:` key in the YAML file; Snakemake doesn't care what the configuration file is called (it can even be a JSON file), but the workflow specifically refers to the `builds` key of the configuration everywhere.

I think of a "build" as @tsibley describes as the output of a "workflow + parameters". "Output" is a useful synonym for "build", except it isn't specific enough to the type of output we produce (a "build" consists of the Auspice JSONs and sometimes the intermediate outputs). The **workflow** is the software that produces the "output" and the **parameters** (e.g., input files, build name, region, country, subsampling scheme, title, etc.) determine how the output is produced by the workflow. If we think of the ncov workflow as a traditional command line tool (let's call it `ncov`),  it might look like this:

```
./ncov \
  --inputs builds.yaml \
  --output-dir auspice/
```

In this example, the `ncov` binary is the workflow and `builds.yaml` are the input parameters. The output JSON files in `auspice` (e.g., `ncov_global.json`, `ncov_global_tip-frequencies.json`, and `ncov_global_root-sequence.json`) comprise the "builds" produced by the workflow and parameters. In this context, it's useful to have a name to refer to these outputs because the "workflow" or the "parameters" by themselves don't tell you what the outputs are. "Datasets" feels like a better description of the input files (as in, "I've curated my dataset to run through the workflow.") rather than the output files. 

> I just didn't want to feel like we're running in circles to accommodate in our terminology a file name that was chosen in the thick of things in early 2020.

My memory of this time is hazy already, but I do remember a meeting with @emmahodcroft, @rneher, and me where we worked out the basic format of the configuration file. I think the term `builds` seemed most natural because that's what we had (and have) been using to refer to the outputs of our workflows.

Reply:
I wrote (emphasis added):
> Renaming builds.yaml to something else seems like it would be quite disruptive at this point. **Even if we make the change backwards compat, people will be confused when the terminology changes.** Build is used pretty pervasively in the ncov docs now.
> …
> Renaming the `builds:` key is maybe less disruptive than the file, but really not sure about that.

@huddlej wrote:
> Renaming `builds.yaml` to anything else is much less disruptive than renaming the `builds:` key in the YAML file; Snakemake doesn't care what the configuration file is called (it can even be a JSON file), but the workflow specifically refers to the `builds` key of the configuration everywhere.

The disruption I was focusing on wasn't technical but social (see emphasis I added above).

Technically, it seems straightforward/mechanical to me to update the workflow to reference `outputs:` and then add logic early in the ncov Snakefile to handle backwards compat with any `builds:` key (e.g. copy it to `outputs:`). So this seems not much harder than updating all our `builds.yaml` references to something else.

Again though, I'm not sure we _should_ do this for ncov.